core/src/main/scala/org/http4s/jdkhttpclient/JdkWSClient.scala

```diff
@@ -47,115 +48,134 @@ object JdkWSClient {
-              isOutputOpen <- F.delay(!webSocket.isOutputClosed)
-              closeOutput = F.fromCompletableFuture(
-                F.delay(webSocket.sendClose(JWebSocket.NORMAL_CLOSURE, ""))
-              )
-              _ <-
-                closeOutput
-                  .whenA(isOutputOpen)
-                  .recover { case e: IOException if e.getMessage == "closed output" => () }
-                  .onError { case e: IOException =>
-                    for {
-                      errs <- Stream
-                        .repeatEval(queue.tryTake)
-                        .unNoneTerminate
-                        .collect { case Left(e) => e }
-                        .compile
-                        .toList
-                      _ <- F.raiseError[Unit](CompositeFailure.fromList(errs) match {
-                        case Some(cf) => cf
-                        case None => e
-                      })
-                    } yield ()
-                  }
-            } yield ()
-          }
-          .map { case (webSocket, queue, closedDef, sendSem) =>
-            // sending will throw if done in parallel
-            val rawSend = (wsf: WSFrame) =>
-              F.fromCompletableFuture(F.delay(wsf match {
-                case WSFrame.Text(text, last) => webSocket.sendText(text, last)
-                case WSFrame.Binary(data, last) => webSocket.sendBinary(data.toByteBuffer, last)
-                case WSFrame.Ping(data) => webSocket.sendPing(data.toByteBuffer)
-                case WSFrame.Pong(data) => webSocket.sendPong(data.toByteBuffer)
-                case WSFrame.Close(statusCode, reason) => webSocket.sendClose(statusCode, reason)
-              }))
-                .void
-            new WSConnection[F] {
-              override def send(wsf: WSFrame) =
-                sendSem.permit.use(_ => rawSend(wsf))
-              override def sendMany[G[_]: Foldable, A <: WSFrame](wsfs: G[A]) =
-                sendSem.permit.use(_ => wsfs.traverse_(rawSend))
-              override def receive = closedDef.tryGet.flatMap {
-                case None => F.delay(webSocket.request(1)) *> queue.take.rethrow.map(_.some)
-                case Some(()) => none[WSFrame].pure[F]
+
+              F.guarantee(cleanupF, ensureResourceReleaseF)
+            }
+            .map { case (webSocket, queue, closedDef, sendSem) =>
+              // sending will throw if done in parallel
+              val rawSend = (wsf: WSFrame) =>
+                F.fromCompletableFuture(F.delay(wsf match {
+                  case WSFrame.Text(text, last) => webSocket.sendText(text, last)
+                  case WSFrame.Binary(data, last) => webSocket.sendBinary(data.toByteBuffer, last)
+                  case WSFrame.Ping(data) => webSocket.sendPing(data.toByteBuffer)
+                  case WSFrame.Pong(data) => webSocket.sendPong(data.toByteBuffer)
+                  case WSFrame.Close(statusCode, reason) => webSocket.sendClose(statusCode, reason)
```

Comment:
One thing I have seen is this line failing because the output was already closed. We don't send close frames manually so the only way this should be invoked is by `connectHighLevel` which mirrors the close frame. Maybe we should guard this with an if, checking if the output is still open? I don't like that much though because low level usage would then have implicit restrictions. 

Also as a side note, the mirroring of the closure frame in `connectHighLevel` actually may fail if the server emits a close code that is not applicable for a client to send.